File: task4/GO/main.go
package main

import (
	"fmt"
)

// getWinner определяет победителя: 1 если Павел, 0 если Вика
func getWinner(n int, m int, prefixSums []int64) int {
	var scorePavel int64 = 0
	var scoreVika int64 = 0
	currentPos := 0
	lastMove := -1
	isPavelTurn := true

	for currentPos < n {
		// Используем очень маленькое число для инициализации поиска максимума
		var maxStepSum int64 = -2e18
		chosenCount := 0

		for i := 1; i <= m; i++ {
			// Проверка на выход за границы и ограничение Варианта 2
			if currentPos+i > n {
				break
			}
			if i == lastMove {
				continue
			}

			// Вычисляем сумму за текущий ход
			currentStepSum := prefixSums[currentPos+i] - prefixSums[currentPos]

			// Жадный выбор: ищем максимальную сумму
			if currentStepSum > maxStepSum {
				maxStepSum = currentStepSum
				chosenCount = i
			} else if currentStepSum == maxStepSum && chosenCount == 0 {
				// Если суммы равны, выбираем минимальное количество чисел
				chosenCount = i
			}
		}

		// Защита от тупиковой ситуации (если легальный ход не найден)
		if chosenCount == 0 {
			if lastMove == 1 {
				chosenCount = 2
			} else {
				chosenCount = 1
			}
			if currentPos+chosenCount > n {
				chosenCount = n - currentPos
			}
			maxStepSum = prefixSums[currentPos+chosenCount] - prefixSums[currentPos]
		}

		// Начисление очков
		if isPavelTurn {
			scorePavel += maxStepSum
		} else {
			scoreVika += maxStepSum
		}

		// Смена состояния
		currentPos += chosenCount
		lastMove = chosenCount
		isPavelTurn = !isPavelTurn
	}

	if scorePavel > scoreVika {
		return 1
	}
	return 0
}

func main() {
	var n, m int
	_, err := fmt.Scan(&n, &m)
	if err != nil {
		return
	}

	// Срез для префиксных сумм (аналог vector в C++)
	prefixSums := make([]int64, n+1)
	for i := 0; i < n; i++ {
		var val int64
		fmt.Scan(&val)
		prefixSums[i+1] = prefixSums[i] + val
	}

	fmt.Println(getWinner(n, m, prefixSums))
}

File: task4/CPP/main.cpp
#include <iostream>
#include <vector>

using namespace std;

// Функция для определения победителя (возвращает 1, если победил Павел, и 0, если Вика)
int getWinner(int n, int m, const vector<long long>& prefixSums) {
    long long scorePavel = 0;
    long long scoreVika = 0;
    int currentPos = 0;
    int lastMove = -1; // Количество чисел, взятых на предыдущем ходу
    bool isPavelTurn = true;

    while (currentPos < n) {
        long long maxStepSum = -2e18; // Минимально возможное значение для поиска максимума
        int chosenCount = 0;

        // Игрок выбирает i чисел (от 1 до m)
        for (int i = 1; i <= m; i++) {
            // Нельзя выйти за пределы последовательности
            if (currentPos + i > n) break;
            
            // Вариант 2: нельзя брать столько же, сколько взял противник только что
            if (i == lastMove) continue;

            // Считаем сумму за этот ход через префиксные суммы
            long long currentStepSum = prefixSums[currentPos + i] - prefixSums[currentPos];

            // Жадная стратегия: ищем максимальную сумму
            if (currentStepSum > maxStepSum) {
                maxStepSum = currentStepSum;
                chosenCount = i;
            } 
            // Если суммы одинаковы, выбираем минимальное количество чисел
            else if (currentStepSum == maxStepSum && chosenCount == 0) {
                chosenCount = i;
            }
        }

        // Если из-за ограничений хода нет, берем минимум (защита от зацикливания)
        if (chosenCount == 0) {
            chosenCount = (lastMove == 1) ? 2 : 1;
            if (currentPos + chosenCount > n) chosenCount = n - currentPos;
            maxStepSum = prefixSums[currentPos + chosenCount] - prefixSums[currentPos];
        }

        // Прибавляем очки текущему игроку
        if (isPavelTurn) {
            scorePavel += maxStepSum;
        } else {
            scoreVika += maxStepSum;
        }

        // Обновляем позицию в массиве и передаем ход
        currentPos += chosenCount;
        lastMove = chosenCount;
        isPavelTurn = !isPavelTurn;
    }

    // Возвращаем результат сравнения счетов
    if (scorePavel > scoreVika) return 1;
    else return 0;
}

int main() {
    int n, m;
    if (!(cin >> n >> m)) return 0;

    // Используем префиксные суммы для быстрого нахождения суммы на отрезке
    vector<long long> prefixSums(n + 1, 0);
    for (int i = 0; i < n; i++) {
        long long val;
        cin >> val;
        prefixSums[i + 1] = prefixSums[i] + val;
    }

    // Вызываем функцию и выводим результат
    cout << getWinner(n, m, prefixSums) << endl;

    return 0;
}